### ISSUE TITLE
Break in template called from loop creates strange results

### ISSUE THREAD

**Issue by PMunch** [nim-lang/Nim#10193]
A template doesn't normally allow a break statement to be passed into a typed block even if the body is inserted into a loop. This makes sense since the break statement doesn't refer to anything the user can see when calling the template. However if such a template is called from a loop and the user adds a break statement that is intended to break the loop the template is called from the compiler doesn't complain and the break statement is added. This however means that it breaks the wrong loop, which can cause some strange issues.

### Example
```nim
template testLoop(x: typed): typed =
  for i in 0..<3:
    x
    
# Prints three "Hello world", as expected
testLoop:
  echo "Hello world"
  
# Illegal to use break here as the template takes a typed block
when false:
  echo "-------"
  testLoop:
    echo "Hello world"
    break

# Would expect this to print one hello world, but prints ten.
echo "-------"
for i in 0..<10:
  testLoop:
    echo "Hello world"
    break
```

### Possible Solution
Adding a block around the outer for loop and breaking that obviously works, but Nim should see the inner break statement as belonging to this loop in the first place and break the correct loop.

**Comment by PMunch** [comment]
This is still an issue

**Comment by Araq** [comment]
> Would expect this to print one hello world, but prints ten.

Well, that's not how template expansion works though.

**Comment by Clyybber** [comment]
Maybe we should convert/bind unnamed breaks into named ones in sem.

**Comment by disruptek** [comment]
> Maybe we should convert/bind unnamed breaks into named ones in sem.

I like this idea, but the current behavior is not at all surprising.

**Comment by PMunch** [comment]
Well, it's not surprising if you know the implementation of the template. But if you're using a library that supplies a template and it happens to use a block for example:
``` Nim
template example(i, x: untyped): untyped =
  block:
    let y {.inject.} = (i * 2) mod 3 # Some silly math
    x # Run our body
    y # This effectively returns y
    

for i in 0..2:
  let z = example i:
    echo "Hello: ", y # Using y directly within the template
    if y == 2:
      break # Would expect this to break the loop, but it just breaks the `block` in the template and messes up the return value
  echo z # Using the y from the template as z
  echo declared(y) # Double checking that y is no longer in our scope
  echo "" # Newline for prettier prints
```
If that `example` template was in a library and you saw that for loop in your code you'd probably assume that the `break` there would stop the loop at the second iteration. But since the template uses a block it will interfere with it and probably lead to unforeseen consequences.

**Comment by Araq** [comment]
Yes but ``block`` inside templates should be rare, instead we have hygienic templates. And if you write a public API don't export templates that have ``block`` inside them, in fact, avoid to export templates altogether.
